pkg/schemapatcher/gen.go

```diff
@@ -356,13 +384,75 @@ func crdsFromDirectory(ctx *genall.GenerationContext, dir string) (map[schema.Gr
 			continue
 		}
 
-		res[groupKind] = &partialCRD{
-			GroupKind:   groupKind,
-			Yaml:        &yamlNodeTree,
-			Versions:    versions,
-			FileName:    fileInfo.Name(),
-			NewSchemata: make(map[string]apiext.JSONSchemaProps),
+		res[*groupKind] = &partialCRD{
+			GroupKind:          *groupKind,
+			Yaml:               &yamlNodeTree,
+			Versions:           versions,
+			FileName:           fileInfo.Name(),
+			NewSchemata:        make(map[string]apiextv1b1.JSONSchemaProps),
+			ValidateStructural: validateStructural,
+			WriteGlobalSchema:  writeGlobalSchema,
 		}
 	}
 	return res, nil
 }
+
+// unmarshalCRDV1 unmarshalls a v1 CRD to the GroupKind kind and a
+// map whose keys identify the versions supported by the crd.
+func unmarshalCRDV1(rawContent []byte) (*schema.GroupKind, map[string]struct{}) {
+	var actualCRD apiextv1.CustomResourceDefinition
+	if err := kyaml.Unmarshal(rawContent, &actualCRD); err != nil {
+		return nil, nil
+	}
+	groupKind := &schema.GroupKind{Group: actualCRD.Spec.Group, Kind: actualCRD.Spec.Names.Kind}
+	var versions map[string]struct{}
+	// CRD v1 lacks a global schema
+	versions = make(map[string]struct{}, len(actualCRD.Spec.Versions))
+	for _, ver := range actualCRD.Spec.Versions {
+		versions[ver.Name] = struct{}{}
+	}
+	return groupKind, versions
+}
+
+// unmarshalCRDV1 unmarshalls a v1beta1 CRD to the GroupKind kind and
+// a map whose keys identify the versions supported by the crd.
+func unmarshalCRDV1beta1(rawContent []byte) (*schema.GroupKind, map[string]struct{}) {
+	var actualCRD apiextv1b1.CustomResourceDefinition
+	if err := kyaml.Unmarshal(rawContent, &actualCRD); err != nil {
+		return nil, nil
+	}
+	groupKind := &schema.GroupKind{Group: actualCRD.Spec.Group, Kind: actualCRD.Spec.Names.Kind}
+	var versions map[string]struct{}
+	if len(actualCRD.Spec.Versions) == 0 {
+		versions = map[string]struct{}{actualCRD.Spec.Version: struct{}{}}
+	} else {
+		versions = make(map[string]struct{}, len(actualCRD.Spec.Versions))
+		for _, ver := range actualCRD.Spec.Versions {
+			versions[ver.Name] = struct{}{}
+		}
+	}
+	return groupKind, versions
+}
+
+// validateStructural checks whether the given openapi schema is
+// structurally valid (required for CRD v1).
+func validateStructural(schema *apiextv1b1.JSONSchemaProps) error {
+	marshalledSchema, err := json.Marshal(schema)
+	if err != nil {
+		return err
+	}
+	genericSchema := &apiext.JSONSchemaProps{}
```

Comment:
@sttts This check for a structural schema is using internal methods (as per discussion on #323). Is there an alternate way to validate that a schema is structural, or is that capability intentionally limited to the k/k repo?

Reply:
like all other validations in kube they are supposed to be internal. 

Before calling NewStructural, you are supposed to validate with apiextensions-apiserver/pkg/apis/apiextensions/validation. That's of course tricky as you don't have a CRD here yet.

tl/dr: I am still not convinced we should validate anything here about the schema. We should rather make schemas structural by construction medium term. Any failing validation here would be a controller-tools bug, not a user error.

Reply:
The latest iteration removes the validation.